src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -0,0 +1,236 @@
+      int createdClientId,
+      Map<String, String> connectorConfig,
+      SnowflakeStreamingIngestClient client) {
+    this.createdClientId = new AtomicInteger(createdClientId);
+    this.connectorConfig = connectorConfig;
+    this.streamingIngestClient = client;
+  }
+
+  /**
+   * Creates a new streaming client, will replace the existing one if necessary
+   *
+   * @param connectorConfig The connector config to define the client
+   */
+  public void createOrReplaceClient(Map<String, String> connectorConfig) {
+    // replace previous connector config and client if applicable
+    this.clientLock.lock();
+    LOGGER.info("Creating new client, this will replace the old client and config if exists");
+    this.closeClient();
+
+    this.connectorConfig = connectorConfig;
+    this.streamingIngestClient = this.initStreamingClient(this.connectorConfig);
+    this.clientLock.unlock();
+  }
+
+  /** Closes the current client */
+  public void closeClient() {
+    // don't do anything if client is already invalid
+    if (!isClientValid(this.streamingIngestClient)) {
+      LOGGER.info("Streaming client already closed");
+      return;
+    }
+
+    this.clientLock.lock();
+    LOGGER.info("Closing Streaming Client:{}", this.streamingIngestClient.getName());
+    try {
+      this.streamingIngestClient.close();
+    } catch (Exception e) {
+      // the client should auto close, so don't throw an exception here
+      String message =
+          e.getMessage() == null || e.getMessage().isEmpty()
+              ? "missing exception message"
+              : e.getMessage();
+      String cause =
+          e.getCause() == null || e.getCause().getStackTrace() == null
+              ? "missing exception cause"
+              : Arrays.toString(e.getCause().getStackTrace());
+
+      LOGGER.error("Failure closing Streaming client msg:{}, cause:{}", message, cause);
+    } finally {
+      this.clientLock.unlock();
+    }
+  }
+
+  /**
+   * Gets the current client or creates a new one from the given connector config If client
+   * optimization is not enabled, just create a new streaming client
+   *
+   * @param connectorConfig The connector config, given as a backup in case the current client is
+   *     invalid
+   * @return The current or newly created streaming client
+   */
+  public SnowflakeStreamingIngestClient getClient(Map<String, String> connectorConfig) {
+    if (Boolean.parseBoolean(
+        connectorConfig.get(
+            SnowflakeSinkConnectorConfig.ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG))) {
+      this.clientLock.lock();
+
+      // recreate streaming client if needed
+      if (!isClientValid(this.streamingIngestClient)) {
+        LOGGER.error("Current streaming client is invalid, recreating client");
+        this.createOrReplaceClient(connectorConfig);
+      }
+
+      this.clientLock.unlock();
+
+      return this.streamingIngestClient;
+    } else {
+      return this.initStreamingClient(connectorConfig);
```

Comment:
```suggestion
     this.streamingIngestClient = this.initStreamingClient(connectorConfig);
     return this.streamingIngestClient;
```

Reply:
Umm okay, ignore the suggestion, I dont think you can do this since this will have a side effect of getting a stale instance. (if you ever introduce a `getStreamingIngestClient`)